### ISSUE TITLE
Unexpected capture of EngineLayer when capturing int value

### ISSUE THREAD

**Issue by gaaclarke** [dart-lang/sdk#50593]
## observed behavior

The following code executes fine:
```dart
  @override
  void initState() {
    super.initState();
    _getAgent().then((agent) async {
      final String? decoded =
          await agent.query((state) => state.decodedMessages[0]);
      if (mounted) {
        setState(() {
          _text = decoded;
        });
      }
    });
  }
```

If I change the integer literal to `index`, the app errors:
```dart
  @override
  void initState() {
    super.initState();
    final int index = widget.index;
    _getAgent().then((agent) async {
      final String? decoded =
          await agent.query((state) => state.decodedMessages[index]);
      if (mounted) {
        setState(() {
          _text = decoded;
        });
      }
    });
  }
```

## error log
```
[VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: Invalid argument(s): Illegal argument in isolate message: (object extends NativeWrapper - Library:'dart:ui' Class: EngineLayer)
#0      _SendPort._sendInternal (dart:isolate-patch/isolate_patch.dart:249:43)
#1      _SendPort.send (dart:isolate-patch/isolate_patch.dart:230:5)
#2      Agent.query (package:isolate_agents/isolate_agents.dart:104:15)
#3      _MessageState.initState.<anonymous closure> (package:romeo_juliet/main.dart:144:23)
<asynchronous suspension>
```

## expected behavior
Since I am copying the `widget.index` value to its own integer, I expect that to get captured, not the widget and thus no error about capturing and sending ui elements to an Isolate.

**Comment by mraleph** [comment]
Duplicate of #36983

**Comment by mraleph** [comment]
tldr: Dart VM creates contexts from scopes, rather than creating a context pre-closure. This makes closures that share the same scope overcapture things, e.g. 

```dart
void f() {
  var x, y;
  foo() => x;
  bar() => y;
}
// Both [foo] and [bar] capture x & y.
```

**Comment by gaaclarke** [comment]
I moved the crux of this issue over to that one since I think this is a novel problem for a known issue.
